File: reference/domain-logic/partname.ts
import type { Category } from "./types";

/**
 * Split a warframe.market display name into a bold set/frame name + a part sub.
 *   "Mesa Prime Systems"        → { name: "Mesa Prime",  sub: "Systems" }
 *   "Nova Prime Chassis Blueprint" → { name: "Nova Prime", sub: "Chassis Blueprint" }
 *   "Saryn Prime Set"           → { name: "Saryn Prime",  sub: "Set" }
 * Non-prime names fall back to the whole name + the catalog part_type.
 */
export function splitName(
  displayName: string,
  partType: string,
): { name: string; sub: string } {
  const i = displayName.indexOf(" Prime");
  if (i >= 0) {
    const name = displayName.slice(0, i + " Prime".length);
    const sub = displayName.slice(i + " Prime".length).trim() || partType;
    return { name, sub };
  }
  return { name: displayName, sub: partType };
}

const FRAME_PARTS = new Set(["Systems", "Chassis", "Neuroptics"]);
const WEAPON_PARTS = new Set([
  "Blade",
  "Blades",
  "Handle",
  "Barrel",
  "Receiver",
  "Stock",
  "String",
  "Link",
  "Pouch",
  "Disc",
  "Lower limb",
  "Upper limb",
  "Head",
  "Grip",
  "Ornament",
  "Boot",
  "Hilt",
  "Gauntlet",
  "Buckle",
  "Guard",
]);

/** Best-effort Warframe/Weapon classification from the catalog part_type. */
export function categoryFor(partType: string): Category {
  if (FRAME_PARTS.has(partType)) return "Warframe";
  if (WEAPON_PARTS.has(partType)) return "Weapon";
  return "Other";
}

File: src-tauri/src/domain/partname.rs
//! Split a warframe.market display name into a bold set/frame name + a part sub.
//!   "Mesa Prime Systems"           → ("Mesa Prime", "Systems")
//!   "Nova Prime Chassis Blueprint" → ("Nova Prime", "Chassis Blueprint")
//!   "Saryn Prime Set"              → ("Saryn Prime", "Set")
//! Non-prime names fall back to (whole name, part_type).

pub fn split_name(display_name: &str, part_type: &str) -> (String, String) {
    if let Some(i) = display_name.find(" Prime") {
        let split = i + " Prime".len();
        let name = display_name[..split].to_string();
        let sub = display_name[split..].trim();
        let sub = if sub.is_empty() {
            part_type.to_string()
        } else {
            sub.to_string()
        };
        (name, sub)
    } else {
        (display_name.to_string(), part_type.to_string())
    }
}
